File: amplitude-integration/server/amplitude_client.py
"""
=====================================================
AMPLITUDE PYTHON CLIENT
=====================================================

Centralized Amplitude SDK wrapper for Python backend/server applications.
Provides a simple API for tracking events from FastAPI, Flask, Django, etc.

Installation:
    pip install amplitude-analytics

Environment Variables Required:
    AMPLITUDE_API_KEY - Your Amplitude project API key

Usage:

    from amplitude_integration.server.amplitude_client import AmplitudeClient
    
    # Initialize once in your app
    client = AmplitudeClient()
    
    # Track events
    client.track(
        user_id='user-123',
        event_type='PRODUCT_DETECTED',
        event_properties={
            'product_id': 'water_bottle',
            'confidence': 0.95,
            'detection_time_ms': 150
        }
    )
    
    # Set user properties
    client.identify(
        user_id='user-123',
        user_properties={
            'device_type': 'kinect_camera',
            'location': 'store_001'
        }
    )
"""

import os
import logging
from typing import Dict, Optional, Any
from datetime import datetime

# Try to import Amplitude SDK
try:
    from amplitude import Amplitude, BaseEvent, EventOptions, Identify
    AMPLITUDE_AVAILABLE = True
except ImportError:
    AMPLITUDE_AVAILABLE = False
    logging.warning(
        "Amplitude SDK not installed. Run: pip install amplitude-analytics"
    )

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class AmplitudeClient:
    """
    Singleton wrapper for Amplitude Python SDK
    """
    
    _instance = None
    
    def __new__(cls, api_key: Optional[str] = None):
        """Singleton pattern - only one instance per process"""
        if cls._instance is None:
            cls._instance = super(AmplitudeClient, cls).__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize Amplitude client
        
        Args:
            api_key: Amplitude API key (optional, reads from AMPLITUDE_API_KEY env var)
        """
        if self._initialized:
            return
        
        # Get API key from parameter or environment
        self.api_key = api_key or os.getenv('AMPLITUDE_API_KEY', '')
        
        if not self.api_key:
            logger.error(
                "AMPLITUDE_API_KEY not configured! Set it in your .env file or pass to AmplitudeClient()"
            )
            self.client = None
            self._initialized = True
            return
        
        # Check if SDK is available
        if not AMPLITUDE_AVAILABLE:
            logger.error("Amplitude SDK not available. Events will not be tracked.")
            self.client = None
            self._initialized = True
            return
        
        # Initialize Amplitude SDK
        try:
            self.client = Amplitude(self.api_key)
            logger.info("✓ Amplitude Python client initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize Amplitude: {e}")
            self.client = None
        
        self._initialized = True
    
    def is_configured(self) -> bool:
        """Check if Amplitude is properly configured"""
        return self.client is not None
    
    def track(
        self,
        user_id: str,
        event_type: str,
        event_properties: Optional[Dict[str, Any]] = None,
        device_id: Optional[str] = None,
        time: Optional[int] = None,
        insert_id: Optional[str] = None,
        **kwargs
    ) -> bool:
        """
        Track an event to Amplitude
        
        Args:
            user_id: User identifier
            event_type: Event name (e.g., 'PRODUCT_DETECTED')
            event_properties: Dictionary of event properties
            device_id: Device identifier (optional)
            time: Event timestamp in milliseconds (optional, defaults to now)
            insert_id: Unique event ID for deduplication (optional)
            **kwargs: Additional event options
            
        Returns:
            bool: True if event was tracked successfully
        """
        if not self.is_configured():
            logger.warning(f"Cannot track event '{event_type}' - Amplitude not configured")
            return False
        
        try:
            # Create event
            event = BaseEvent(
                event_type=event_type,
                user_id=user_id,
                device_id=device_id,
                event_properties=event_properties or {},
                time=time,
                insert_id=insert_id
            )
            
            # Track event
            self.client.track(event)
            
            logger.debug(f"Tracked event: {event_type} for user {user_id}")
            return True
            
        except Exception as e:
            logger.error(f"Error tracking event '{event_type}': {e}")
            return False
    
    def identify(
        self,
        user_id: str,
        user_properties: Dict[str, Any],
        device_id: Optional[str] = None
    ) -> bool:
        """
        Set or update user properties
        
        Args:
            user_id: User identifier
            user_properties: Dictionary of user properties to set
            device_id: Device identifier (optional)
            
        Returns:
            bool: True if identification was successful
        """
        if not self.is_configured():
            logger.warning(f"Cannot identify user '{user_id}' - Amplitude not configured")
            return False
        
        try:
            # Create identify event
            identify_obj = Identify()
            
            # Set user properties
            for key, value in user_properties.items():
                identify_obj.set(key, value)
            
            # Create event with identify
            event = BaseEvent(
                event_type="$identify",
                user_id=user_id,
                device_id=device_id,
                user_properties=user_properties
            )
            
            # Track identify event
            self.client.track(event)
            
            logger.debug(f"Identified user: {user_id} with properties: {list(user_properties.keys())}")
            return True
            
        except Exception as e:
            logger.error(f"Error identifying user '{user_id}': {e}")
            return False
    
    def flush(self) -> bool:
        """
        Flush all pending events to Amplitude
        
        Returns:
            bool: True if flush was successful
        """
        if not self.is_configured():
            return False
        
        try:
            self.client.flush()
            logger.debug("Flushed events to Amplitude")
            return True
        except Exception as e:
            logger.error(f"Error flushing events: {e}")
            return False
    
    def shutdown(self):
        """
        Shutdown the Amplitude client and flush remaining events
        Call this on application shutdown
        """
        if not self.is_configured():
            return
        
        try:
            self.client.shutdown()
            logger.info("Amplitude client shutdown complete")
        except Exception as e:
            logger.error(f"Error during shutdown: {e}")


# Singleton instance for easy importing
amplitude_client = AmplitudeClient()


# Convenience functions that use the singleton
def track_event(
    user_id: str,
    event_type: str,
    event_properties: Optional[Dict[str, Any]] = None,
    **kwargs
) -> bool:
    """
    Convenience function to track an event
    Uses the singleton AmplitudeClient instance
    """
    return amplitude_client.track(user_id, event_type, event_properties, **kwargs)


def identify_user(
    user_id: str,
    user_properties: Dict[str, Any],
    device_id: Optional[str] = None
) -> bool:
    """
    Convenience function to identify a user
    Uses the singleton AmplitudeClient instance
    """
    return amplitude_client.identify(user_id, user_properties, device_id)


def flush_events() -> bool:
    """
    Convenience function to flush events
    Uses the singleton AmplitudeClient instance
    """
    return amplitude_client.flush()


# Example usage for detection events from your backend
def track_product_detection(
    user_id: str,
    product_id: str,
    confidence: float,
    detection_time_ms: int,
    camera_id: str = 'kinect_001'
):
    """
    Helper function to track product detection events
    
    Args:
        user_id: User/session identifier
        product_id: Detected product ID
        confidence: Detection confidence score
        detection_time_ms: Time taken for detection
        camera_id: Camera/sensor identifier
    """
    return track_event(
        user_id=user_id,
        event_type='PRODUCT_DETECTED',
        event_properties={
            'product_id': product_id,
            'confidence': confidence,
            'detection_time_ms': detection_time_ms,
            'camera_id': camera_id,
            'timestamp': datetime.utcnow().isoformat()
        }
    )


def track_grounding_dino_inference(
    user_id: str,
    classes_detected: list,
    num_detections: int,
    inference_time_ms: int,
    model_name: str = 'grounding-dino-tiny'
):
    """
    Helper function to track Grounding DINO inference events
    
    Args:
        user_id: User/session identifier
        classes_detected: List of detected class names
        num_detections: Number of detections
        inference_time_ms: Inference time in milliseconds
        model_name: Model identifier
    """
    return track_event(
        user_id=user_id,
        event_type='MODEL_INFERENCE',
        event_properties={
            'model_name': model_name,
            'classes_detected': classes_detected,
            'num_detections': num_detections,
            'inference_time_ms': inference_time_ms,
            'timestamp': datetime.utcnow().isoformat()
        }
    )

File: amplitude-integration/browser/amplitude-client.js
/**
 * =====================================================
 * AMPLITUDE BROWSER CLIENT
 * =====================================================
 * 
 * Centralized Amplitude SDK initialization for browser/frontend apps.
 * Provides a simple API for initializing and using Amplitude across
 * multiple frontend applications.
 * 
 * Usage:
 * 
 * // In your HTML:
 * <script type="module">
 *   import { AmplitudeClient } from './amplitude-integration/browser/amplitude-client.js';
 *   
 *   // Initialize once in your app
 *   AmplitudeClient.init({
 *     apiKey: 'your-api-key',
 *     userId: 'user-123' // optional
 *   });
 *   
 *   // Track events anywhere
 *   AmplitudeClient.track('BUTTON_CLICKED', { buttonName: 'signup' });
 * </script>
 * 
 * Or use the hooks for specific features:
 * - ecommerce-hooks.js: Product tracking
 * - dashboard-hooks.js: Analytics dashboard tracking
 * - experiment-hooks.js: A/B testing & feature flags
 */

import { AmplitudeConfig } from '../config/amplitude.config.js';

/**
 * AmplitudeClient - Singleton wrapper for Amplitude Browser SDK
 */
class AmplitudeBrowserClient {
  constructor() {
    this.isInitialized = false;
    this.amplitude = null;
    this.userId = null;
    this.sessionId = null;
  }

  /**
   * Initialize Amplitude SDK
   * 
   * @param {Object} options - Initialization options
   * @param {string} options.apiKey - Amplitude API key (optional, reads from config)
   * @param {string} options.userId - User identifier (optional)
   * @param {Object} options.userProperties - Initial user properties (optional)
   * @param {Object} options.sdkOptions - Override SDK options (optional)
   * @returns {Promise<boolean>} - Returns true if initialization successful
   */
  async init(options = {}) {
    if (this.isInitialized) {
      console.warn('[AmplitudeClient] Already initialized');
      return true;
    }

    // Check if Amplitude SDK is loaded
    if (typeof window === 'undefined' || !window.amplitude) {
      console.error('[AmplitudeClient] Amplitude SDK not found! Make sure to include the Amplitude script in your HTML.');
      console.error('Add this to your <head>: <script src="https://cdn.amplitude.com/libs/analytics-browser-2.3.8-min.js.gz"></script>');
      return false;
    }

    // Get API key from options or config
    const apiKey = options.apiKey || AmplitudeConfig.getApiKey();
    
    if (!apiKey) {
      console.error('[AmplitudeClient] No API key provided. Set AMPLITUDE_API_KEY in .env or pass apiKey to init()');
      return false;
    }

    try {
      // Initialize the SDK
      this.amplitude = window.amplitude;
      
      // Merge options with defaults
      const sdkOptions = {
        ...AmplitudeConfig.options,
        ...(options.sdkOptions || {})
      };

      // Initialize
      await this.amplitude.init(apiKey, options.userId, sdkOptions);
      
      this.isInitialized = true;
      this.userId = options.userId || null;

      console.log('[AmplitudeClient] ✓ Initialized successfully', {
        project: AmplitudeConfig.projectName,
        userId: this.userId,
        sessionId: await this.getSessionId()
      });

      // Set initial user properties if provided
      if (options.userProperties) {
        this.setUserProperties(options.userProperties);
      }

      return true;
    } catch (error) {
      console.error('[AmplitudeClient] Initialization failed:', error);
      return false;
    }
  }

  /**
   * Track an event
   * 
   * @param {string} eventName - Event name (use AmplitudeConfig.events constants)
   * @param {Object} eventProperties - Event properties
   */
  track(eventName, eventProperties = {}) {
    if (!this.isInitialized || !this.amplitude) {
      console.warn('[AmplitudeClient] Not initialized. Call init() first.');
      return;
    }

    try {
      this.amplitude.track(eventName, eventProperties);
      console.log(`[AmplitudeClient] Tracked: ${eventName}`, eventProperties);
    } catch (error) {
      console.error('[AmplitudeClient] Track error:', error);
    }
  }

  /**
   * Identify a user
   * 
   * @param {string} userId - User identifier
   * @param {Object} userProperties - User properties (optional)
   */
  identify(userId, userProperties = {}) {
    if (!this.isInitialized || !this.amplitude) {
      console.warn('[AmplitudeClient] Not initialized. Call init() first.');
      return;
    }

    try {
      this.userId = userId;
      
      const identifyEvent = new this.amplitude.Identify();
      
      // Set user properties
      Object.entries(userProperties).forEach(([key, value]) => {
        identifyEvent.set(key, value);
      });

      this.amplitude.setUserId(userId);
      this.amplitude.identify(identifyEvent);
      
      console.log(`[AmplitudeClient] Identified user: ${userId}`, userProperties);
    } catch (error) {
      console.error('[AmplitudeClient] Identify error:', error);
    }
  }

  /**
   * Set user properties
   * 
   * @param {Object} properties - User properties to set
   */
  setUserProperties(properties) {
    if (!this.isInitialized || !this.amplitude) {
      console.warn('[AmplitudeClient] Not initialized. Call init() first.');
      return;
    }

    try {
      const identifyEvent = new this.amplitude.Identify();
      
      Object.entries(properties).forEach(([key, value]) => {
        identifyEvent.set(key, value);
      });

      this.amplitude.identify(identifyEvent);
      console.log('[AmplitudeClient] User properties set:', properties);
    } catch (error) {
      console.error('[AmplitudeClient] Set user properties error:', error);
    }
  }

  /**
   * Get current session ID
   * 
   * @returns {Promise<number|null>} - Session ID
   */
  async getSessionId() {
    if (!this.isInitialized || !this.amplitude) {
      return null;
    }

    try {
      const session = await this.amplitude.getSessionId();
      this.sessionId = session;
      return session;
    } catch (error) {
      console.error('[AmplitudeClient] Get session ID error:', error);
      return null;
    }
  }

  /**
   * Get current device ID
   * 
   * @returns {Promise<string|null>} - Device ID
   */
  async getDeviceId() {
    if (!this.isInitialized || !this.amplitude) {
      return null;
    }

    try {
      return await this.amplitude.getDeviceId();
    } catch (error) {
      console.error('[AmplitudeClient] Get device ID error:', error);
      return null;
    }
  }

  /**
   * Flush events immediately (useful before page unload)
   */
  async flush() {
    if (!this.isInitialized || !this.amplitude) {
      return;
    }

    try {
      await this.amplitude.flush();
      console.log('[AmplitudeClient] Events flushed');
    } catch (error) {
      console.error('[AmplitudeClient] Flush error:', error);
    }
  }

  /**
   * Reset user & device (e.g., on logout)
   */
  reset() {
    if (!this.isInitialized || !this.amplitude) {
      return;
    }

    try {
      this.amplitude.reset();
      this.userId = null;
      this.sessionId = null;
      console.log('[AmplitudeClient] Reset complete');
    } catch (error) {
      console.error('[AmplitudeClient] Reset error:', error);
    }
  }

  /**
   * Get the raw Amplitude instance (for advanced usage)
   * 
   * @returns {Object|null} - Amplitude SDK instance
   */
  getAmplitudeInstance() {
    return this.amplitude;
  }

  /**
   * Check if initialized
   * 
   * @returns {boolean}
   */
  isReady() {
    return this.isInitialized && this.amplitude !== null;
  }
}

// Export singleton instance
export const AmplitudeClient = new AmplitudeBrowserClient();

// Export the class itself for advanced usage
export { AmplitudeBrowserClient };

// Export config for convenience
export { AmplitudeConfig };

// Make available globally (for non-module scripts)
if (typeof window !== 'undefined') {
  window.AmplitudeClient = AmplitudeClient;
  window.AmplitudeConfig = AmplitudeConfig;
}

export default AmplitudeClient;
